### PR TITLE
Add semi-structured sparse tutorial

### PR DESCRIPTION
**Pull request by jcaip**
- Add new semi-structured sparsity tutorial.
- update index.rst
- add image

### CODE REVIEW

**Comment by cpuhrsch** [comment]
https://docs-preview.pytorch.org/pytorch/tutorials/2574/prototype/semi_structured_sparse.html?highlight=accelelerating%20bert%20semi%20structured is the link to the rendered tutorial

File: prototype_source/semi_structured_sparse.rst [hunk 1]
```diff
@@ -0,0 +1,542 @@
...
+
+These sparse kerenels work by allowing us to skip calculations involving pruned elements. To do this, these kerenls work on sparse matrices, which are do not store the pruned elements and store the specified elements in a compressed format.
+For semi-structured sparsity, we store exactly half of the original parameters along with some compressed metadata about how the elements were arranged.
+
+.. image:: https://developer-blogs.nvidia.com/wp-content/uploads/2023/06/2-4-structured-sparsity-pattern.png
+    :align: center
+    :width: 80%
+
+
+However, this is not the only way to store sparse tensors. There exist other formats like `COO <https://pytorch.org/docs/2.1/sparse.html#sparse-coo-tensors>`_ representation, which is used with **unstructured sparsity**.
```
Comment (svekars) [inline prototype_source/semi_structured_sparse.rst:84]:
```suggestion
However, this is not the only way to store sparse tensors. There are other formats like `COO <https://pytorch.org/docs/2.1/sparse.html#sparse-coo-tensors>`_ representation, which are used with **unstructured sparsity**.
```

File: prototype_source/semi_structured_sparse.rst [hunk 2]
```diff
@@ -0,0 +1,542 @@
...
+    +--------------------+----------------+--------------+---------------------+
+
+
+Conclusion
+----------
+In this tutorial, we have shown how to prune BERT to be 2:4 sparse and how to accelerate a 2:4 sparse model for inference.
+By taking advantage of our SparseSemiStructuredTensor subclass, we were able to achieve a 1.3x speedup over the fp16 baseline.
+We also demonstrated the benefits of 2:4 sparsity by fine-tuning BERT to recover any lost F1. (86.92 dense vs 86.48 sparse)
+
+Please feel free to raise an `issue <https://github.com/pytorch/pytorch/issues>`_ if you need any support, and thanks for reading
```
Comment (svekars) [inline prototype_source/semi_structured_sparse.rst:542]:
```suggestion
```
The rest of the conclusion is great!

File: prototype_source/semi_structured_sparse.rst [hunk 3]
```diff
@@ -0,0 +1,542 @@
...
+    +--------------------+----------------+--------------+---------------------+
+    | Time (bs=256)      | 286.91         | 244.19       | 1.18x               |
+    +--------------------+----------------+--------------+---------------------+
+
+
+Conclusion
+----------
+In this tutorial, we have shown how to prune BERT to be 2:4 sparse and how to accelerate a 2:4 sparse model for inference.
+By taking advantage of our SparseSemiStructuredTensor subclass, we were able to achieve a 1.3x speedup over the fp16 baseline.
+We also demonstrated the benefits of 2:4 sparsity by fine-tuning BERT to recover any lost F1. (86.92 dense vs 86.48 sparse)
```
Comment (svekars) [inline prototype_source/semi_structured_sparse.rst:540]:
```suggestion
We also demonstrated the benefits of 2:4 sparsity by fine-tuning BERT to recover any lost F1 (86.92 dense vs 86.48 sparse).
```

**Review by svekars — COMMENTED** [review]
A few minor editorial nits. Thank you!

File: prototype_source/semi_structured_sparse.rst [hunk 4]
```diff
@@ -0,0 +1,542 @@
+(prototype) Accelelerating BERT with semi-structured (2:4) sparsity
+================================================================
+**Author**: `Jesse Cai <https://github.com/jcaip>`_
+
+Like other forms of sparsity, **semi-structured sparsity** is a model optimization technique that seeks to reduce the memory overhead and latency of a neural network at the expense of some model accuracy.
+It is also known as **fine-grained structured sparsity** or **2:4 structured sparsity**.
+
+Semi-structured sparsity derives its name from its unique sparsity pattern, where n out of every 2n elements are pruned.
+In practice, we most often see n = 2, hence 2:4 sparsity. This sparsity pattern is particularly interesting for two reasons:
```
Comment (jisaacso) [inline prototype_source/semi_structured_sparse.rst:9]: I would collapse this down to something like

Today, we most often see n = 2, hence 2:4 sparsity. 2:4 semi-structured sparsity was designed to be efficiently accelerated on GPUs. In 2021 (?confirm?), NVIDIA introduced hardware support for semi-structured sparsity in their Ampere architecture, and have also released fast sparse kernels via CUTLASS/`cuSPARSELt <https://docs.nvidia.com/cuda/cusparselt/index.html>`_.

The second point isn't saying much unless you go into much more detail on alternative sparsity methods, and why they can compress more.

File: prototype_source/semi_structured_sparse.rst [hunk 5]
```diff
@@ -0,0 +1,542 @@
...
+
+        import torch
+        from torch.sparse import to_sparse_semi_structured, SparseSemiStructuredTensor
+        from torch.utils.benchmark import Timer
+        SparseSemiStructuredTensor._FORCE_CUTLASS = True
+
+        # mask Linear weight to be 2:4 sparse
+        mask = torch.Tensor([0, 0, 1, 1]).tile((3072, 2560)).cuda().bool()
+        linear = torch.nn.Linear(10240, 3072).half().cuda().eval()
+        linear.weight = torch.nn.Parameter(mask * linear.weight)
```
Comment (jisaacso) [inline prototype_source/semi_structured_sparse.rst:44]: why do you need to mask out weights on the dense model?

File: prototype_source/semi_structured_sparse.rst [hunk 6]
```diff
@@ -0,0 +1,542 @@
...
+
+    On my machine (A100 80GB), I see: `Dense: 0.870ms Sparse: 0.630ms | Speedup: 1.382x`
+
+
+What problem does semi-structured sparsity solve?
+-------------------------------------------------
+The general motivation behind sparsity is simple: if there are zeros in your network, you can avoid storing / doing compute with those parameters.
+However, the specifics of sparsity are tricky. Zeroing out parameters doesn't affect the latency / memory overhead of our model out of the box.
+
+This is because the dense tensor itself still contains the pruned elements and will still operate on those elements during matrix multiplication.
```
Comment (jisaacso) [inline prototype_source/semi_structured_sparse.rst:74]: introduce the word kernel here. 

"This is because the dense tensor itself still contains the pruned elements, and the respective matrix multiplication kernel will still operate on those elements."

File: prototype_source/semi_structured_sparse.rst [hunk 7]
```diff
@@ -0,0 +1,542 @@
...
+In order to realize performance gains, we need to swap out our dense kernels for sparse kernels.
```
Comment (jisaacso) [inline prototype_source/semi_structured_sparse.rst:75]: "our" -> "a"

File: prototype_source/semi_structured_sparse.rst [hunk 8]
```diff
@@ -0,0 +1,542 @@
...
+
+These sparse kerenels work by allowing us to skip calculations involving pruned elements. To do this, these kerenls work on sparse matrices, which are do not store the pruned elements and store the specified elements in a compressed format.
```
Comment (jisaacso) [inline prototype_source/semi_structured_sparse.rst:77]: strike "by allowing us" ... "work by skipping calculations...". Also kernels is spelled wrong.

File: prototype_source/semi_structured_sparse.rst [hunk 9]
```diff
@@ -0,0 +1,542 @@
...
+For semi-structured sparsity, we store exactly half of the original parameters along with some compressed metadata about how the elements were arranged.
+
+.. image:: https://developer-blogs.nvidia.com/wp-content/uploads/2023/06/2-4-structured-sparsity-pattern.png
```
Comment (jisaacso) [inline prototype_source/semi_structured_sparse.rst:80]: if you're going to embed this image, give credit in a caption from where it was taken.

File: prototype_source/semi_structured_sparse.rst [hunk 10]
```diff
@@ -0,0 +1,542 @@
...
+    :align: center
+    :width: 80%
+
+
+However, this is not the only way to store sparse tensors. There are other formats like `COO <https://pytorch.org/docs/2.1/sparse.html#sparse-coo-tensors>`_ representation, which are used with **unstructured sparsity**.
```
Comment (jisaacso) [inline prototype_source/semi_structured_sparse.rst:85]: If you're going to include this blurb on COO, I'd move it up top into the intro where you talk about alternative sparsity methods. I don't think it makes sense here. You're just getting into the flow of showcasing the API

**Review by jisaacso — COMMENTED** [review]
First pass without running tutorial code, just feedback on prose. Second pass later.

File: index.rst [hunk 11]
```diff
@@ -544,6 +544,13 @@ What's new in PyTorch tutorials?
    :link: advanced/static_quantization_tutorial.html
    :tags: Quantization
 
+.. customcarditem::
```
Comment (svekars) [inline index.rst:547]: please add to the https://github.com/pytorch/tutorials/blob/main/prototype_source/prototype_index.rst rather than in index.rst

File: prototype_source/semi_structured_sparse.rst (hunk 4 above)
Comment (jcaip) [reply prototype_source/semi_structured_sparse.rst:9]: I think it's important to mention that 2:4 sparsity degrades model accuracy less than other sparse methods. 

I'm thinking something like:
```
Today, we most often see n = 2, hence 2:4 sparsity. Semi-structured (2:4) sparsity is particularly interesting because it offers GPU hardware acceleration with minimal loss to model accuracy. In 2020, NVIDIA introduced hardware support for semi-structured sparsity in their Ampere architecture, and have also released fast sparse kernels via CUTLASS/cuSPARSELt <https://docs.nvidia.com/cuda/cusparselt/index.html>_. 
```

File: prototype_source/semi_structured_sparse.rst (hunk 10 above)
Comment (jcaip) [reply prototype_source/semi_structured_sparse.rst:85]: I want to frame semi-structured sparsity as a kind of "goldilocks" between unstructured and structured sparsity, but I agree that this is kind of abrupt.

**Review by jisaacso — APPROVED** [review]
Thanks for addressing comments, looks great
